File: lib/core/reader/reader_document.dart
/// A parsed reader document, either EPUB or PDF.
sealed class ReaderDocument {
  String get filePath;
  String? get initialLocator;
}

class EpubDocument extends ReaderDocument {
  @override
  final String filePath;
  @override
  final String? initialLocator;
  final String htmlContent;
  final int chapterCount;

  EpubDocument({
    required this.filePath,
    this.initialLocator,
    required this.htmlContent,
    required this.chapterCount,
  });

  EpubDocument copyWith({String? initialLocator}) {
    return EpubDocument(
      filePath: filePath,
      initialLocator: initialLocator ?? this.initialLocator,
      htmlContent: htmlContent,
      chapterCount: chapterCount,
    );
  }
}

class PdfDocument extends ReaderDocument {
  @override
  final String filePath;
  @override
  final String? initialLocator;
  final int initialPageIndex;

  PdfDocument({
    required this.filePath,
    this.initialLocator,
    required this.initialPageIndex,
  });

  PdfDocument copyWith({String? initialLocator, int? initialPageIndex}) {
    return PdfDocument(
      filePath: filePath,
      initialLocator: initialLocator ?? this.initialLocator,
      initialPageIndex: initialPageIndex ?? this.initialPageIndex,
    );
  }
}

File: core/reader/src/main/kotlin/com/readler/core/reader/ReaderDocument.kt
package com.readler.core.reader

sealed class ReaderDocument {
    abstract val filePath: String
    abstract val initialLocator: String?

    data class Epub(
        override val filePath: String,
        override val initialLocator: String?,
        val htmlContent: String,
        val chapterCount: Int
    ) : ReaderDocument()

    data class Pdf(
        override val filePath: String,
        override val initialLocator: String?,
        val initialPageIndex: Int
    ) : ReaderDocument()
}
